pkg/controller/operators/catalog/step.go

```diff
@@ -0,0 +1,264 @@
+package catalog
+
+import (
+	"encoding/json"
+	"fmt"
+	"k8s.io/client-go/dynamic"
+
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	crdlib "github.com/operator-framework/operator-lifecycle-manager/pkg/lib/crd"
+	index "github.com/operator-framework/operator-lifecycle-manager/pkg/lib/index"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	errorwrap "github.com/pkg/errors"
+	logger "github.com/sirupsen/logrus"
+	v1ext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	v1beta1ext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Stepper manages cluster interactions based on the step.
+type Stepper interface {
+	Status() (v1alpha1.StepStatus, error)
+}
+
+// StepperFunc fulfills the Stepper interface.
+type StepperFunc func() (v1alpha1.StepStatus, error)
+
+func (s StepperFunc) Status() (v1alpha1.StepStatus, error) {
+	return s()
+}
+
+// Builder holds clients and data structures required for the StepBuilder to work
+// Builder attributes are not to meant to be accessed outside the StepBuilder method
+type builder struct {
+	opclient          operatorclient.ClientInterface
+	dynamicClient     dynamic.Interface
+	csvToProvidedAPIs map[string]cache.Indexer
+}
+
+func newBuilder(opclient operatorclient.ClientInterface, dynamicClient dynamic.Interface, csvToProvidedAPIs map[string]cache.Indexer) *builder {
+	return &builder{
+		opclient:          opclient,
+		dynamicClient:     dynamicClient,
+		csvToProvidedAPIs: csvToProvidedAPIs,
+	}
+}
+
+type notSupportedStepperErr struct {
+	message string
+}
+
+func (n notSupportedStepperErr) Error() string {
+	return n.message
+}
+
+// step is a factory that creates StepperFuncs based on the Kind provided and the install plan step.
+func (b *builder) step(stepStatus v1alpha1.StepStatus, manifest, name string) (Stepper, error) {
+	gvk := metav1.GroupVersionKind{}
+	err := json.Unmarshal([]byte(manifest), &gvk)
+	if err != nil {
+		return nil, errorwrap.Wrapf(err, "error finding the %s GVK", name)
+	}
+
+	switch gvk.Kind {
+	case crdKind:
+		switch gvk.Version {
+		case crdlib.V1Beta1Version:
```

Comment:
> If I understand you're saying we can GET v1 CRDs even though the manifest was v1beta1
yup

>  I can try deduplicating it - was hesitant to change the existing flow too much but maybe I'm just being overly cautious.
The reason I brought this up is because the two functions are nearly identical.

Reply:
I spoke with Evan and it seems like we want to convert to V1 or the internal version first from the manifest.

Reply:
Right that makes sense - I'm arriving at the same conclusion

Reply:
Ok so my last commit does a conversion between v1beta1 and v1 via scheme.Convert - now the two codepaths are identical with the exception of the serialization and the `Create` call with the v1beta1 client - I guess next step is to consolidate them even further

Reply:
I agree, just trying to see how to best do so